### PR TITLE
Clean up Dyn

### PR DESCRIPTION
**Pull request by whitequark**
**Kind:** cleanup/documentation.

This PR has no functional change. 

While working on https://github.com/coq/coq/pull/7850#issuecomment-398415569, I found that the Dyn module is not documented and the code is rather hard to follow with inconsistent indentation and opaque names.

I documented it and made it hopefully slightly easier to follow. This PR is best viewed [with no whitespace changes](https://github.com/coq/coq/pull/7880/files?w=1).

### CODE REVIEW

**Review by ejgallego — APPROVED** [review]
LGTM, thanks

**Comment by ejgallego** [comment]
One warning slipped tru tho.

**Comment by whitequark** [comment]
Fixed.

**Review by ppedrot — APPROVED** [review]

**Comment by whitequark** [comment]
It's green!
